physics/physics.js

```diff
@@ -0,0 +1,505 @@
+      [ "rgba(255,53,53,1)", "rgba(181,41,41,1)" ],
+      [ "rgba(8,160,0,1)",   "rgba(24,102,23,1)" ],
+      [ "rgba(219,219,0,1)", "rgba(178,173,41,1)" ]
+    ];
+    var top = 100;
+    var left = 200;
+
+    var numSpheres = 80;
+    for (var i = 0; i < numSpheres; ++i) {
+      var radius = widths[i % widths.length];
+      var colorPair = colors[i % colors.length];
+      var noise = i % 2 == 0 ? -5 : 5;
+      var dx = left - radius + noise;
+      var dy = top - radius * 2;
+
+      createSphere(radius, colorPair[0], colorPair[1], dx, dy);
+
+      top -= (widths[i % widths.length] * 2) + 30;
+      if (i == Math.floor(numSpheres/3)) {
+        top = 500;
+        left = 650;
+      } else if (i == Math.floor((numSpheres*2)/3)) {
+        top = 200;
+        left = 950;
+      }
+    }
+  }
+
+  function setupFoils() {
+    var dxs = [ 50, 500, 700 ];
+    var dys = [ 150, 600, 300 ];
+    var delays = [ 0, 500, 250 ];
+    for (var i = 0; i < 3; ++i)
+      foils.push(createFoil(dxs[i], dys[i], 300, 30, delays[i]));
+  }
+
+  function setupScene() {
+    logo = document.getElementById("logo");
+    viewportHeight = document.documentElement.clientHeight;
+
+    var container = document.querySelector(".container");
+    container.addEventListener("mousedown", function(e) { e.preventDefault(); });
+    container.addEventListener("click", toggle);
+
+    setupFoils();
+    setupSpheres();
+  }
+
+  scope.initWorker = function() {
+    self.onmessage = function(e) {
+      spheres = e.data.spheres;
+      foils = e.data.foils;
+      viewportHeight = e.data.viewportHeight;
+      requestAnimationFrame(tick);
+    }
+  };
+
+  scope.initMain = function() {
+    setupScene();
+    scope.worker = new CompositorWorker("physics.js");
+
+    var message = {
+      'spheres': [],
+      'foils': [],
+      'viewportHeight': viewportHeight
+    };
+
+    for (var i = 0; i < spheres.length; ++i) {
+      var sphere = spheres[i];
+      var sphereProxy = {
+        radius: sphere.radius,
+        vx: sphere.vx,
+        vy: sphere.vy,
+        dx: sphere.dx,
+        dy: sphere.dy,
+        initialX: sphere.initialX,
+        initialY: sphere.initialY,
+        mass:sphere.mass,
+        blurElements: []
+      };
+      for (var j = 0; j < sphere.blurElements.length; ++j)
+        sphereProxy.blurElements.push(new CompositorProxy(sphere.blurElements[j], ['transform']));
+      message.spheres.push(sphereProxy);
+    }
+
+    for (var i = 0; i < foils.length; ++i) {
+      var foil = foils[i];
+      var foilProxy = {
+        left: foil.left,
+        top: foil.top,
+        width:foil.width,
+        height:foil.height,
+        dx:foil.dx,
+        dy:foil.dy,
+        vx:foil.vx,
+        vy:foil.vy,
+        actor:new CompositorProxy(foil.actor, ['transform'])
+      }
+      message.foils.push(foilProxy);
+    }
+
+    worker.postMessage(message);
+    requestAnimationFrame(tick);
+  };
+
+  if (isMain())
```

Comment:
For educational purposes, I’d prefer splitting main thread JS and worker JS into two files. This is just an additional barrier when trying to grok this thing.